### ISSUE TITLE
STM32F303 I2C `blocking_read_write` works but async `write_read` fails due to a timeout

### ISSUE THREAD

**Issue by fmckeogh** [embassy-rs/embassy#2884]
`main.rs`:

```
bind_interrupts!(struct Irqs {
    I2C1_EV => i2c::EventInterruptHandler<peripherals::I2C1>;
    I2C1_ER => i2c::ErrorInterruptHandler<peripherals::I2C1>;
});

#[embassy_executor::main]
async fn main(spawner: Spawner) {
    let p = embassy_stm32::init(Config::default());

    let mut config = i2c::Config::default();

    let mut i2c = I2c::new(
        p.I2C1,
        p.PB6,
        p.PB7,
        Irqs,
        p.DMA1_CH2,
        p.DMA1_CH7,
        Hertz(100_000),
        Default::default(),
    );

    let mut buf = [0u8];
    i2c.blocking_write_read(0x38, &[0u8], &mut buf).unwrap();

    defmt::info!("blocking: {:?}", buf);

    let mut buf = [0u8];
    i2c.write_read(0x38, &[0u8], &mut buf).await.unwrap();

    defmt::info!("async: {:?}", buf);
}
```

Output:

```
INFO  blocking: [24]
└─ pd_interceptor::____embassy_main_task::{async_fn#0} @ pd-interceptor/src/main.rs:55  
ERROR panicked at pd-interceptor/src/main.rs:58:50:
called `Result::unwrap()` on an `Err` value: Timeout
```

Why does the async read-write fail? Have I set up the interrupts/DMA correctly?

**Comment by jamesmunns** [comment]
@fmckeogh I looked and the F303 doesn't have the errata from #2887, I would suggest adding logs to the `poll_fn`s in the read/write dma internal to figure out which branch was taken, I'm guessing in the hanging case, it's taking one of the branches that DOESN'T re-enable the interrupts.

Since the interrupts self-disable, any non-terminating branch of the poll-fns should re-enable the interrupts.

**Comment by fmckeogh** [comment]
@jamesmunns Awesome, will do:)

**Comment by fmckeogh** [comment]
@jamesmunns Yes, seems like you're right. The contents of the `poll_fn` gets called once, returns pending, then is never called again. I've tried manually re-enabling interrupts but no luck so far, I think I need to take some time to understand the codebase.

**Comment by EzraWolf** [comment]
Do you have any updates?

**Comment by fmckeogh** [comment]
@EzraWolf Sorry, my last comment is as far as I got with the debugging.

**Comment by ProfFan** [comment]
The I2C v2 driver is still broken at this moment. Likely the async implementation is incorrect.

**Comment by yodaldevoid** [comment]
I'm currently running into this sporadically. It appears to be timing related as when I enable more logging, and consequently slow down the system, the problem disappears.

Right now it looks like the call to `I2c::master_write` in `I2c::write_dma_internal` sometimes gets stuck waiting, locking up the I2C `write_read` process, and causing the timeout. I'm having trouble reproducing it once I add logging in `I2c::master_write` so I cannot currently determine if it is stuck waiting for the address to write out or the bus to be free.

**Comment by yodaldevoid** [comment]
I've finally been able to trigger the issue, though it still is a little unreliable. What definitely is happening is a write is started, gets stuck waiting for the bus to not be busy, and then times out.

~~Based on a few logic analyzer captures and some choice debug logs it appears that the device holding the bus busy is the controller itself. As far as I can tell `embassy_embedded_hal::shared_bus::asynch::i2c::I2cDevice` is not properly sharing the I2C bus and is allowing device drivers to interrupt each other. Said another way, one device driver starts a transaction, a second device driver butts in and starts a write, the second device driver then sees that the bus is busy and waits until the timeout, and if you don't panic at this point control eventually is passed back to the first device driver to complete the transaction.~~

EDIT: On second thought, it doesn't make sense for the problem to be at the `I2cDevice` level. What is definitely happening is that the first transaction is getting cut off before it is completed. This is before the stop bit and for reads before the final NACK. I have also seen a read get cut off before ACKing the requested number of bytes. At this point the first transaction is considered completed and the second transaction is started. This second transaction waits for the bus to be ready, the bus is never freed by the controller, and the timeout occurs.

**Comment by ProfFan** [comment]
It's a race
